File: java/src/test/java/com/notifyservice/WebhooksTest.java
package com.notifyservice;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Base64;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import org.junit.jupiter.api.Test;

class WebhooksTest {

    private static final byte[] RAW = "super-secret-bytes-0123456789abcd".getBytes(StandardCharsets.UTF_8);

    private static String encodedSecret() {
        return "whsec_" + Base64.getUrlEncoder().withoutPadding().encodeToString(RAW);
    }

    private static String sign(byte[] raw, long ts, String body) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(raw, "HmacSHA256"));
            byte[] sig = mac.doFinal((ts + "." + body).getBytes(StandardCharsets.UTF_8));
            StringBuilder hex = new StringBuilder();
            for (byte b : sig) {
                hex.append(String.format("%02x", b));
            }
            return "t=" + ts + ",v1=" + hex;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Test
    void validSignaturePasses() {
        String body = "{\"event\":\"notification.delivered\"}";
        long ts = Instant.now().getEpochSecond();
        assertTrue(Webhooks.verifySignature(body, sign(RAW, ts, body), encodedSecret()));
    }

    @Test
    void tamperedBodyFails() {
        String header = sign(RAW, Instant.now().getEpochSecond(), "{\"event\":\"a\"}");
        assertFalse(Webhooks.verifySignature("{\"event\":\"b\"}", header, encodedSecret()));
    }

    @Test
    void expiredTimestampFails() {
        long ts = Instant.now().getEpochSecond() - 10_000;
        assertFalse(Webhooks.verifySignature("{}", sign(RAW, ts, "{}"), encodedSecret()));
    }

    @Test
    void wrongSecretFails() {
        long ts = Instant.now().getEpochSecond();
        String header = sign(RAW, ts, "{}");
        String other = "whsec_" + Base64.getUrlEncoder().withoutPadding()
                .encodeToString("different-secret-bytes-aaaaaaaaaa".getBytes(StandardCharsets.UTF_8));
        assertFalse(Webhooks.verifySignature("{}", header, other));
    }

    @Test
    void malformedHeaderFails() {
        assertFalse(Webhooks.verifySignature("{}", "garbage", encodedSecret()));
    }

    @Test
    void secretWithoutPrefixWorks() {
        long ts = Instant.now().getEpochSecond();
        String bare = Base64.getUrlEncoder().withoutPadding().encodeToString(RAW);
        assertTrue(Webhooks.verifySignature("{}", sign(RAW, ts, "{}"), bare));
    }
}

File: php/tests/WebhooksTest.php
<?php

declare(strict_types=1);

namespace NotifyService\Tests;

use NotifyService\Webhooks;
use PHPUnit\Framework\TestCase;

final class WebhooksTest extends TestCase
{
    /** @return array{0: string, 1: string} encoded secret + raw bytes */
    private function secret(): array
    {
        $raw = 'super-secret-bytes-0123456789abcd';
        $encoded = 'whsec_' . rtrim(strtr(base64_encode($raw), '+/', '-_'), '=');
        return [$encoded, $raw];
    }

    private function sign(string $raw, int $ts, string $body): string
    {
        $mac = hash_hmac('sha256', $ts . '.' . $body, $raw);
        return "t={$ts},v1={$mac}";
    }

    public function testValidSignaturePasses(): void
    {
        [$secret, $raw] = $this->secret();
        $body = '{"event":"notification.delivered"}';
        $ts = time();
        self::assertTrue(Webhooks::verifySignature($body, $this->sign($raw, $ts, $body), $secret));
    }

    public function testTamperedBodyFails(): void
    {
        [$secret, $raw] = $this->secret();
        $header = $this->sign($raw, time(), '{"event":"a"}');
        self::assertFalse(Webhooks::verifySignature('{"event":"b"}', $header, $secret));
    }

    public function testExpiredTimestampFails(): void
    {
        [$secret, $raw] = $this->secret();
        $ts = time() - 10000;
        self::assertFalse(Webhooks::verifySignature('{}', $this->sign($raw, $ts, '{}'), $secret));
    }

    public function testWrongSecretFails(): void
    {
        [, $raw] = $this->secret();
        $ts = time();
        $header = $this->sign($raw, $ts, '{}');
        $other = 'whsec_' . rtrim(strtr(base64_encode('different-secret-bytes-aaaaaaaaaa'), '+/', '-_'), '=');
        self::assertFalse(Webhooks::verifySignature('{}', $header, $other));
    }

    public function testMalformedHeaderFails(): void
    {
        [$secret, ] = $this->secret();
        self::assertFalse(Webhooks::verifySignature('{}', 'not-a-valid-header', $secret));
    }

    public function testSecretWithoutPrefixWorks(): void
    {
        [, $raw] = $this->secret();
        $ts = time();
        $bare = rtrim(strtr(base64_encode($raw), '+/', '-_'), '=');
        self::assertTrue(Webhooks::verifySignature('{}', $this->sign($raw, $ts, '{}'), $bare));
    }
}
